File: Session3/Java/BinarySearch.java
import java.util.Scanner;
class BinarySearch{  

    public static int search(int arr[], int target)
    {
        int n = arr.length;
        int s=0;
        int e=n-1;
        while(s<=e) {
            int mid = (s+e)/2;
            // int mid = s+(e-s)/2;
            if(arr[mid]==target) {
                return mid;
            }
            else if(arr[mid] < target) {
                s = mid+1;
            }
            else {
                e = mid-1;
            }
        }
        return -1;
    }

    public static void main(String args[]){  
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int target = sc.nextInt();
        int[] arr = new int[n];
        for(int i=0; i < arr.length; i++)  {
            arr[i] = sc.nextInt();
        }
        int result = search(arr, target);
        System.out.print(result);
    }  
}  

File: Session3/C++/BinarySearch.cpp
#include<bits/stdc++.h>
using namespace std;

int binarySearch(int arr[], int target, int n)
{
    int s=0;
    int e=n-1; 
    while(s<=e) {
        int mid = (s+e)/2;
        // int mid = s+(e-s)/2; --> integer overflow
        if(arr[mid]==target) {
            return mid;
        }
        else if(arr[mid] < target) {
            s = mid+1; // right
        }
        else {
            e = mid-1; // left
        }
    }
    return -1;
}

int main(){
    int n;
    cin>>n;
    int target;
    cin>>target;
    int arr[n];
    for(int i=0; i<n; i++) {
        cin>>arr[i];
    }
    cout<<binarySearch(arr, target, n);
    return 0;
}